File: src/application/use_cases/discovery.py
"""Discovery-focused application use cases."""

from __future__ import annotations

from collections import Counter
from pathlib import Path
from typing import Callable, Sequence, TypeVar

from src.application.ports import InventoryCatalog
from src.application.use_cases.inventory import list_cluster_targets
from src.domain.discovery import (
    ClientPage,
    ClientSummary,
    HostPage,
    HostQuery,
    HostRecord,
    HostResolutionResult,
    PageInfo,
)
from src.domain.models import ClusterTarget

T = TypeVar("T")


def _optional_str(value: object) -> str | None:
    if value is None:
        return None
    text = str(value).strip()
    return text or None


def _systemframe_id_for(target: ClusterTarget) -> str | None:
    return _optional_str(
        target.host_config.get("systemframe_id")
        or target.group_vars.get("systemframe_id")
    )


def project_target_to_host_record(target: ClusterTarget) -> HostRecord:
    return HostRecord(
        client=target.company,
        host_name=target.host_alias,
        systemframe_id=_systemframe_id_for(target),
        addr_ip=_optional_str(target.host_config.get("ansible_host")),
        context_name=target.context_name,
        group=target.group,
    )


def load_host_records(
    inventory_path: Path,
    catalog: InventoryCatalog,
) -> list[HostRecord]:
    return [
        project_target_to_host_record(target)
        for target in list_cluster_targets(inventory_path, catalog)
    ]


def _paginate(
    items: Sequence[T],
    *,
    limit: int,
    cursor: str | None,
    key_fn: Callable[[T], str],
) -> tuple[tuple[T, ...], PageInfo]:
    normalized_limit = max(1, limit)
    ordered = sorted(items, key=key_fn)
    start_index = 0

    if cursor is not None:
        for index, item in enumerate(ordered):
            if key_fn(item) > cursor:
                start_index = index
                break
        else:
            start_index = len(ordered)

    selected = tuple(ordered[start_index : start_index + normalized_limit])
    has_more = start_index + normalized_limit < len(ordered)

    return selected, PageInfo(
        limit=normalized_limit,
        returned=len(selected),
        total=len(ordered),
        has_more=has_more,
        next_cursor=key_fn(selected[-1]) if has_more and selected else None,
    )


def list_client_summaries(
    inventory_path: Path,
    catalog: InventoryCatalog,
    *,
    query: str | None = None,
    limit: int = 20,
    cursor: str | None = None,
) -> ClientPage:
    counts = Counter(record.client for record in load_host_records(inventory_path, catalog))
    items = [
        ClientSummary(client=client, host_count=host_count)
        for client, host_count in counts.items()
        if query is None or query.lower() in client.lower()
    ]
    page_items, page = _paginate(
        items,
        limit=limit,
        cursor=cursor,
        key_fn=lambda item: item.client,
    )
    return ClientPage(items=page_items, page=page)


def _matches_value(value: str | None, expected: str | None, *, exact: bool) -> bool:
    if expected is None:
        return True
    if value is None:
        return False

    left = value.lower()
    right = expected.lower()
    if exact:
        return left == right
    return left.startswith(right) or right in left


def _matches_query(record: HostRecord, query: HostQuery) -> bool:
    if query.client is not None and record.client != query.client:
        return False
    if not _matches_value(record.host_name, query.host_name, exact=query.exact):
        return False
    if not _matches_value(record.systemframe_id, query.systemframe_id, exact=query.exact):
        return False
    if not _matches_value(record.addr_ip, query.addr_ip, exact=query.exact):
        return False
    if not _matches_value(record.context_name, query.context_name, exact=True):
        return False

    if query.query is None:
        return True

    search_text = query.query.lower()
    haystacks = (
        record.host_name.lower(),
        record.client.lower(),
        record.context_name.lower(),
        (record.systemframe_id or "").lower(),
        (record.addr_ip or "").lower(),
    )
    return any(search_text in haystack for haystack in haystacks)


def search_hosts(
    inventory_path: Path,
    catalog: InventoryCatalog,
    *,
    query: HostQuery,
    limit: int = 20,
    cursor: str | None = None,
) -> HostPage:
    matches = [
        record
        for record in load_host_records(inventory_path, catalog)
        if _matches_query(record, query)
    ]
    page_items, page = _paginate(
        matches,
        limit=limit,
        cursor=cursor,
        key_fn=lambda item: item.context_name,
    )
    return HostPage(items=page_items, page=page, query=query)


def resolve_host(
    inventory_path: Path,
    catalog: InventoryCatalog,
    *,
    query: HostQuery,
    limit: int = 10,
) -> HostResolutionResult:
    page = search_hosts(
        inventory_path,
        catalog,
        query=query,
        limit=limit,
        cursor=None,
    )

    if page.page.total == 1:
        record = page.items[0]
        return HostResolutionResult(
            status="unique",
            query=query,
            matches=page.items,
            page=page.page,
            context_name=record.context_name,
            hint=None,
        )

    if page.page.total == 0:
        return HostResolutionResult(
            status="no_match",
            query=query,
            matches=(),
            page=page.page,
            context_name=None,
            hint="No hosts matched the provided identifiers.",
        )

    return HostResolutionResult(
        status="ambiguous",
        query=query,
        matches=page.items,
        page=page.page,
        context_name=None,
        hint="Multiple hosts matched. Refine with --ip, --id, or a more specific host name.",
    )

File: internal/domain/discovery.go
package domain

import (
	"sort"
	"strings"
)

// HostRecord is a read-only view of a single cluster host.
type HostRecord struct {
	Client        string
	HostName      string
	SystemframeID *string
	Addr          *string
	ContextName   string
	Group         string
	Status        *string // peer connectivity status (e.g. "Connected", "Connecting", "Idle")
}

// ClientSummary holds the host count for a client.
type ClientSummary struct {
	Client    string
	HostCount int
}

// PageInfo describes a paginated result set.
type PageInfo struct {
	Limit      int
	Returned   int
	Total      int
	HasMore    bool
	NextCursor *string
}

// HostQuery is a filter for host lookups.
type HostQuery struct {
	Client        *string
	HostName      *string
	SystemframeID *string
	Addr          *string
	ContextName   *string
	Query         *string
	Exact         bool
}

// ClientPage is a paginated list of clients.
type ClientPage struct {
	Items []ClientSummary
	Page  PageInfo
}

// HostPage is a paginated list of hosts.
type HostPage struct {
	Items []HostRecord
	Page  PageInfo
	Query HostQuery
}

// ResolutionStatus describes the outcome of a host resolution attempt.
type ResolutionStatus string

// Resolution outcomes.
const (
	ResolutionUnique    ResolutionStatus = "unique"
	ResolutionNoMatch   ResolutionStatus = "no_match"
	ResolutionAmbiguous ResolutionStatus = "ambiguous"
)

// HostResolutionResult is the structured output of a host resolution query.
type HostResolutionResult struct {
	Status      ResolutionStatus
	Query       HostQuery
	Matches     []HostRecord
	Page        PageInfo
	ContextName *string
	Hint        *string
}

// ProjectTargetToHostRecord converts a ClusterTarget to a HostRecord.
func ProjectTargetToHostRecord(target ClusterTarget) HostRecord {
	hc := target.HostConfig()
	gv := target.GroupVars()

	var sfID, ip, status *string
	if v, ok := hc["systemframe_id"].(string); ok {
		sfID = &v
	} else if v, ok := gv["systemframe_id"].(string); ok {
		sfID = &v
	}
	if v, ok := hc["addr"].(string); ok {
		ip = &v
	}
	if v, ok := hc["netbird_status"].(string); ok {
		status = &v
	}

	return HostRecord{
		Client:        target.Company(),
		HostName:      target.HostAlias(),
		SystemframeID: sfID,
		Addr:          ip,
		ContextName:   target.ContextName(),
		Group:         target.Group(),
		Status:        status,
	}
}

// BuildHostRecords projects a slice of ClusterTargets to HostRecords.
func BuildHostRecords(targets []ClusterTarget) []HostRecord {
	records := make([]HostRecord, len(targets))
	for i, t := range targets {
		records[i] = ProjectTargetToHostRecord(t)
	}
	return records
}

// BuildClientPage groups host records by client and returns a paginated page sorted by client name.
func BuildClientPage(records []HostRecord, limit int, cursor string) ClientPage {
	counts := map[string]int{}
	for _, r := range records {
		counts[r.Client]++
	}

	clients := make([]string, 0, len(counts))
	for c := range counts {
		clients = append(clients, c)
	}
	sort.Strings(clients)

	start := 0
	if cursor != "" {
		for i, c := range clients {
			if c == cursor {
				start = i + 1
				break
			}
		}
	}

	all := clients[start:]
	total := len(counts)

	end := len(all)
	if limit > 0 && limit < end {
		end = limit
	}

	page := all[:end]
	hasMore := end < len(all)

	var nextCursor *string
	if hasMore && len(page) > 0 {
		c := page[len(page)-1]
		nextCursor = &c
	}

	items := make([]ClientSummary, len(page))
	for i, c := range page {
		items[i] = ClientSummary{Client: c, HostCount: counts[c]}
	}

	return ClientPage{
		Items: items,
		Page: PageInfo{
			Limit:      limit,
			Returned:   len(items),
			Total:      total,
			HasMore:    hasMore,
			NextCursor: nextCursor,
		},
	}
}

// SearchHostRecords filters and paginates an in-memory record set.
func SearchHostRecords(records []HostRecord, query HostQuery, limit int, cursor string) HostPage {
	filtered := filterRecords(records, query)
	total := len(filtered)

	start := 0
	if cursor != "" {
		for i, r := range filtered {
			if r.ContextName == cursor {
				start = i + 1
				break
			}
		}
	}

	page := filtered[start:]
	end := len(page)
	if limit > 0 && limit < end {
		end = limit
	}
	page = page[:end]
	hasMore := start+end < total

	var nextCursor *string
	if hasMore && len(page) > 0 {
		c := page[len(page)-1].ContextName
		nextCursor = &c
	}

	return HostPage{
		Items: page,
		Page: PageInfo{
			Limit:      limit,
			Returned:   len(page),
			Total:      total,
			HasMore:    hasMore,
			NextCursor: nextCursor,
		},
		Query: query,
	}
}

// ResolveHostRecords resolves a query against in-memory records, returning a unique/ambiguous/no-match outcome.
func ResolveHostRecords(records []HostRecord, query HostQuery, limit int) HostResolutionResult {
	hostPage := SearchHostRecords(records, query, limit, "")
	matches := hostPage.Items
	total := hostPage.Page.Total

	switch total {
	case 0:
		hint := "No hosts matched the provided identifiers."
		return HostResolutionResult{
			Status:  ResolutionNoMatch,
			Query:   query,
			Matches: []HostRecord{},
			Page:    hostPage.Page,
			Hint:    &hint,
		}
	case 1:
		cn := matches[0].ContextName
		return HostResolutionResult{
			Status:      ResolutionUnique,
			Query:       query,
			Matches:     matches,
			Page:        hostPage.Page,
			ContextName: &cn,
		}
	default:
		hint := "Multiple hosts matched. Refine with --addr, --id, or a more specific host name."
		return HostResolutionResult{
			Status:  ResolutionAmbiguous,
			Query:   query,
			Matches: matches,
			Page:    hostPage.Page,
			Hint:    &hint,
		}
	}
}

func filterRecords(records []HostRecord, q HostQuery) []HostRecord {
	var out []HostRecord
	for _, r := range records {
		if q.Client != nil && r.Client != *q.Client {
			continue
		}
		if q.HostName != nil {
			if q.Exact {
				if r.HostName != *q.HostName {
					continue
				}
			} else if !strings.Contains(r.HostName, *q.HostName) {
				continue
			}
		}
		if q.SystemframeID != nil {
			if r.SystemframeID == nil || *r.SystemframeID != *q.SystemframeID {
				continue
			}
		}
		if q.Addr != nil {
			if r.Addr == nil || *r.Addr != *q.Addr {
				continue
			}
		}
		if q.ContextName != nil && r.ContextName != *q.ContextName {
			continue
		}
		if q.Query != nil {
			needle := *q.Query
			match := strings.Contains(r.HostName, needle) ||
				strings.Contains(r.ContextName, needle) ||
				(r.SystemframeID != nil && strings.Contains(*r.SystemframeID, needle)) ||
				(r.Addr != nil && strings.Contains(*r.Addr, needle))
			if !match {
				continue
			}
		}
		out = append(out, r)
	}
	return out
}
